native/dlaqr4.go

```diff
@@ -0,0 +1,467 @@
+		nwupbd := min(nh, nwmax)
+		if ndfl < kexnw {
+			nw = min(nwupbd, nwr)
+		} else {
+			nw = min(nwupbd, 2*nw)
+		}
+		if nw < nwmax {
+			if nw >= nh-1 {
+				nw = nh
+			} else {
+				kwtop := kbot - nw + 1
+				if math.Abs(h[kwtop*ldh+kwtop-1]) > math.Abs(h[(kwtop-1)*ldh+kwtop-2]) {
+					nw++
+				}
+			}
+		}
+		if ndfl < kexnw {
+			ndec = -1
+		} else if ndec >= 0 || nw >= nwupbd {
+			ndec++
+			if nw-ndec < 2 {
+				ndec = 0
+			}
+			nw -= ndec
+		}
+
+		// Split workspace under the subdiagonal of H into:
+		//  - an nw×nw work array V in the lower left-hand corner,
+		//  - an nw×nho horizontal work array along the bottom edge (nho
+		//    must be at least nw but more is better),
+		//  - an nve×nw vertical work array along the left-hand-edge
+		//    (nve can be any positive integer but more is better).
+		kv := n - nw
+		kt := nw
+		nho := n - (nw + 1) - kt
+		kwv := nw + 1
+		nve := n - nw - kwv
+
+		// Aggressive early deflation.
+		ls, ld := impl.Dlaqr2(wantt, wantz, n, ktop, kbot, nw,
+			h, ldh, iloz, ihiz, z, ldz, wr[:kbot+1], wi[:kbot+1],
+			h[kv*ldh:], ldh, nho, h[kv*ldh+kt:], ldh, nve, h[kwv*ldh:], ldh, work, lwork)
+
+		// Adjust kbot accounting for new deflations.
+		kbot -= ld
+		// ks points to the shifts.
+		ks := kbot - ls + 1
+
+		// Skip an expensive QR sweep if there is a (partly heuristic)
+		// reason to expect that many eigenvalues will deflate without
+		// it. Here, the QR sweep is skipped if many eigenvalues have
+		// just been deflated or if the remaining active block is small.
+		if ld > 0 && (100*ld > nw*nibble || kbot-ktop+1 <= min(nmin, nwmax)) {
+			// ld is positive, note progress.
+			ndfl = 1
+			continue
+		}
+
+		// ns is the nominal number of simultaneous shifts. This may be
+		// lowered (slightly) if Dlaqr2 did not provide that many
+		// shifts.
+		ns := min(min(nsmax, nsr), max(2, kbot-ktop))
+		ns -= ns % 2
+
+		// If there have been no deflations in a multiple of kexsh
+		// iterations, then try exceptional shifts. Otherwise use shifts
+		// provided by Dlaqr2 above or from the eigenvalues of a
+		// trailing principal submatrix.
+		if ndfl%kexsh == 0 {
+			ks = kbot - ns + 1
+			for i := kbot; i > max(ks, ktop+1); i -= 2 {
+				ss := math.Abs(h[i*ldh+i-1]) + math.Abs(h[(i-1)*ldh+i-2])
+				aa := wilk1*ss + h[i*ldh+i]
+				_, _, _, _, wr[i-1], wi[i-1], wr[i], wi[i], _, _ =
+					impl.Dlanv2(aa, ss, wilk2*ss, aa)
+			}
+			if ks == ktop {
+				wr[ks+1] = h[(ks+1)*ldh+ks+1]
+				wi[ks+1] = 0
+				wr[ks] = wr[ks+1]
+				wi[ks] = wi[ks+1]
+			}
+		} else {
+			// If we got ns/2 or fewer shifts, use Dlahqr on a
+			// trailing principal submatrix to get more. Since
+			// ns <= nsmax <=(n+6)/9, there is enough space below
+			// the subdiagonal to fit an ns×ns scratch array.
+			if kbot-ks+1 <= ns/2 {
+				ks = kbot - ns + 1
+				kt = n - ns
+				impl.Dlacpy(blas.All, ns, ns, h[ks*ldh+ks:], ldh, h[kt*ldh:], ldh)
+				inf := impl.Dlahqr(false, false, ns, 0, ns-1, h[kt*ldh:], ldh,
```

Comment:
`inf` here is horrendously misleading. Can we either name it what it actually is, `unconverged`, or just say `ks += impl.Dlahqr(...`?